### ISSUE TITLE
Easier way to add custom attributes

### ISSUE THREAD

**Issue by jishnu7** [tngan/samlify#373]
## Problem:
If we want to add a custom attribute in idp post bindings, the only way is to provide `loginResponseTemplate` string and `customTagReplacement` callback, and inside `customTagReplacement`, we have to replace not just the attributes we want to add, we'll have to replace all the variables in the template, including time conditions, status code etc. 

## Suggested Solution
In `binding-post` we already generate a `tvalue` object which contains all the tag values that are going to be replaced in the template, so pass it as a parameter for `customTagReplacement` function along with the template.

This way, someone implementing customTagReplacement just have to add additional attributes that they are intending to add.

If this suggestion looks good, I can send a PR for this change.

**Comment by Gonyoda** [comment]
I have this need as well.  I was able to apply these changes to base64LoginResponse:

```
    if (idpSetting.loginResponseTemplate) {
      rawSamlResponse = libsaml.replaceTagsByValue(idpSetting.loginResponseTemplate.context, tvalue);

      if (Array.isArray(idpSetting.loginResponseTemplate.attributes)) {
         rawSamlResponse = libsaml.replaceTagsByValue(rawSamlResponse, libsaml.attributeStatementTagBuilder(idpSetting.loginResponseTemplate.attributes, user));
      } else {
        console.log('idpSetting.loginResponseTemplate.attributes is not array');
      }

      if (customTagReplacement) {
         const template =customTagReplacement(rawSamlResponse);
         rawSamlResponse = get(template, 'context', null);
      }

```
and adding to libsaml:

```
   /* @desc Helper function to build the AttributeStatement tag values
    * @param  {LoginResponseAttribute} attributes    an array of attribute configuration
    * @param  {any} user                             The user
    * @return {any}
    */
    attributeStatementTagBuilder(attributes: LoginResponseAttribute[], user: any): any {
      return attributes.reduce((r, { valueTag }) => {
        r[tagging('attr', valueTag)] = user[valueTag.replace('user.', '')];
        return r;
      }, {});

    },
```

This change performs the default replacement, then uses the `user` parameter passed to the base64LoginResponse to replace the additional attributes.

I made additional changes so that the attributes can just be an array of strings, and all the attribute types (eg: valueXsiType) just default to sane values.  This made my IDP code much smaller and easier to understand.

However, fundamentally, this code is flawed (depending on your point of view) because the whole process is just a naive string replacement.  It doesn't account for embedded XML characters (like the less-than or greater-than symbol) or the (low) chance that a replacement could contain the text of another string to replace.  I fear that some clever/naive user could screw up the SAML and maybe cause problems down the line.  (I am no SAML expert, but I'm guessing this would just fail, but I don't know for sure).

I believe that using string replacement is wrong across the board, and instead elements of the SAML XML should be created and combined using objects that are serialized into XML when needed.  This change is clearly beyond the scope of this specific issue and would be a pretty big change to the library from what I can tell.  However, it would make it much more robust, I believe.

**Comment by jimisaacs** [comment]
I need this as well

**Comment by jimisaacs** [comment]
needing to completely duplicate this logic: https://github.com/tngan/samlify/blob/2e1a93671f4b98980472d4857e08a2f99a236acd/src/binding-post.ts#L93-L123

is seriously making me reconsider using this package.

@Gonyoda do you have your changes ready?

**Comment by jimisaacs** [comment]
@Gonyoda I don't see a PR so I may fork and apply them.

**Comment by Gonyoda** [comment]
No, I gave up trying to modify this project due to concerns using it in production and our timeline. Instead, we wrote our client in a .NET Core API with the relatively cheap (for a company anyhow) solution from ComponentSpace.
https://www.componentspace.com/  We just pass the SAML tokens to this .NET API instead of our node API.

I looked for the changes I made but apparently did not keep them - I didn't even fork the repo!  You should be able to make those changes above in your own fork and confirm they work then suggest a fix back to this repo with a PR.

**Comment by jimisaacs** [comment]
Yep, on it, thanks for the the help and the context @Gonyoda

**Comment by jimisaacs** [comment]
I've made some major changes described here: https://github.com/tngan/samlify/issues/411

**Comment by tngan** [comment]
The "build the AttributeStatement for me" shortcut you sketched landed in v2 — `loginResponseTemplate.attributes` is processed at IdP-construction time and the AttributeStatement gets baked into the template before any callback runs. From [`docs/template.md`](https://samlify.js.org/#/template):

```ts
const idp = saml.IdentityProvider({
  loginResponseTemplate: {
    context: '<samlp:Response>...{AttributeStatement}...</samlp:Response>',
    attributes: [
      { name: 'mail', valueTag: 'user.email', nameFormat: 'urn:oasis:names:tc:SAML:2.0:attrname-format:basic', valueXsiType: 'xs:string' },
      { name: 'name', valueTag: 'user.name', nameFormat: 'urn:oasis:names:tc:SAML:2.0:attrname-format:basic', valueXsiType: 'xs:string' },
    ],
  },
});
```

The `{AttributeStatement}` placeholder in your template is replaced with the rendered `<saml:AttributeStatement>` produced from your `attributes` config — you no longer have to hand-write the full attribute markup. Per-call values (`user.email`, etc.) are passed through the standard tag map.

If you specifically want the `customTagReplacement` callback to receive the `tvalue` map so it can do *incremental* edits without re-implementing every replacement, that's a separate, narrower request — feel free to open a fresh issue with that framing if you still want it. The options-bag pattern introduced in #612 / #618 is the right shape for adding a `tags` field to the callback signature in a backwards-compatible way.

Closing as addressed by the v2 attributes shortcut.
